src/lib/io/sol-memmap-storage-impl-zephyr.c

```diff
@@ -0,0 +1,285 @@
+        }
+    }
+
+    if (mask) {
+        for (i = 0, j = 0; i < entry->size; i++, j += 8)
+            value |= (uint64_t)((uint8_t *)buffer->data)[i] << j;
+
+        value &= mask;
+        value >>= entry->bit_offset;
+
+        memset(buffer->data, 0, buffer->capacity);
+        for (i = 0; i < entry->size; i++, value >>= 8)
+            ((uint8_t *)buffer->data)[i] = value & 0xff;
+    }
+
+    buffer->used = entry->size;
+
+    return 0;
+
+err:
+    sol_buffer_fini(buffer);
+    SOL_WRN("Flash read failed");
+    return r;
+}
+
+/* W25QXXDV memories won't write to any memory region that is not
+ * erased, and the minimum erasable block size is ERASE_ALIGN (also
+ * aligned to that size). We have to read back that much bytes before
+ * any write, and we chose the heap to do so. */
+int
+sol_memmap_impl_write_raw(struct map_internal *map_internal,
+    const char *name,
+    const struct sol_memmap_entry *entry,
+    uint64_t mask,
+    struct sol_blob *blob,
+    void (*cb)(void *data, const char *name, struct sol_blob *blob, int status),
+    const void *data)
+{
+    int r = 0;
+    uint8_t *back_mem;
+    size_t offset, mod, back_sz, cnt;
+
+    if (!sol_blob_ref(blob))
+        return -errno;
+
+    offset = FLASH_MEM_REGION_OFFSET + entry->offset;
+    mod = offset % ERASE_ALIGN;
+
+    /* check if we're spamming > 1 erase blocks */
+    back_sz = ERASE_ALIGN * (1 + ((mod + entry->offset) / ERASE_ALIGN));
+
+    back_mem = malloc(back_sz);
+    SOL_NULL_CHECK(back_mem, -ENOMEM);
+
+    for (cnt = 0; cnt < back_sz; cnt += MAX_RW) {
+        r = flash_read(flash_dev, offset - mod + cnt, back_mem + cnt, MAX_RW);
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+    }
+
+    r = flash_write_protection_set(flash_dev, false);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+    r = flash_erase(flash_dev, offset - mod, back_sz);
+    if (r < 0) {
+        SOL_WRN("Flash erase (before write) failed");
+        goto error;
+    }
+
+    if (mask) {
+        uint64_t value = 0, old_value;
+        uint32_t i, j;
+
+        /* entry->size > 8 implies that no mask should be used */
+        assert(entry->size <= 8);
+
+        for (i = 0, j = 0; i < entry->size; i++, j += 8)
+            value |= (uint64_t)((uint8_t *)blob->mem)[i] << j;
+
+        memcpy(&old_value, back_mem + mod, entry->size);
+
+        value <<= entry->bit_offset;
+        value &= mask;
+        value |= (old_value & ~mask);
+
+        memcpy(back_mem + mod, &value, entry->size);
+
+        r = flash_write_protection_set(flash_dev, false);
+        SOL_INT_CHECK_GOTO(r, < 0, error);
+
+        for (cnt = 0; cnt < back_sz; cnt += MAX_RW) {
+            r = flash_write(flash_dev, offset - mod + cnt, back_mem + cnt,
+                MAX_RW);
+            if (r < 0) {
+                SOL_WRN("Flash write failed");
+                goto error;
+            }
+            r = flash_write_protection_set(flash_dev, false);
+            SOL_INT_CHECK_GOTO(r, < 0, error);
```

Comment:
On Mon, 23 May 2016 07:01:23 -0700, Gustavo Lima Chaves wrote:
> > +        value |= (old_value & ~mask);
> > +
> > +        memcpy(back_mem + mod, &value, entry->size);
> > +
> > +        r = flash_write_protection_set(flash_dev, false);
> > +        SOL_INT_CHECK_GOTO(r, < 0, error);
> > +
> > +        for (cnt = 0; cnt < back_sz; cnt += MAX_RW) {
> > +            r = flash_write(flash_dev, offset - mod + cnt, back_mem + cnt,
> > +                MAX_RW);
> > +            if (r < 0) {
> > +                SOL_WRN("Flash write failed");
> > +                goto error;
> > +            }
> > +            r = flash_write_protection_set(flash_dev, false);
> > +            SOL_INT_CHECK_GOTO(r, < 0, error);
> 
> After each write it protects itself again. Deal with it.

The Zephyr API does that? Or is a thing of that particular controller
and the Zephyr API is too close to it instead of trying to abstract
that away?